### PR TITLE
[ANCHOR-462] Junit tests parallelization (core/api-schema)

### PR DESCRIPTION
**Pull request by lijamie98**
### Description

1. Turn on Gradle concurrency for `core` and `api-schema`
2. Remove `clearAllMocks` and `unmockAll` calls in `@AfterEach` functions to avoid side-effects on concurrent tests. 
3. Carefully handle `mockkStatic` objects.
4. `Sep10Challenge` is suspected to be thread-unsafe. Turn on `synchronized` to avoid race condition.

### Context

Test acceleration.

### Testing

`./gradlew test`

### Known limitations

NA.

### CODE REVIEW

File: platform/build.gradle.kts [hunk 1]
```diff
@@ -55,6 +55,9 @@ dependencies {
 }
 
 tasks.test {
+  // Enable parallel test execution
+  systemProperty("junit.jupiter.execution.parallel.enabled", true)
+//  systemProperty("junit.jupiter.execution.parallel.mode.default", "concurrent")
```
Comment (lijamie98) [inline platform/build.gradle.kts:60]: Many of the tests in `platform` subproject are not thread safe.

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt [hunk 2]
```diff
@@ -4,38 +4,34 @@ import io.mockk.*
...
 import org.junit.jupiter.params.provider.EnumSource
-import org.stellar.anchor.auth.ApiAuthJwt.CallbackAuthJwt
-import org.stellar.anchor.auth.ApiAuthJwt.CustodyAuthJwt
-import org.stellar.anchor.auth.ApiAuthJwt.PlatformAuthJwt
+import org.stellar.anchor.auth.ApiAuthJwt.*
 import org.stellar.anchor.auth.AuthType.*
 import org.stellar.anchor.util.AuthHeader
 
+@Execution(SAME_THREAD)
+@Disabled
```
Comment (philipliu) [inline core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt:17]: Should this be disabled?

File: api-schema/build.gradle.kts [hunk 3]
```diff
@@ -13,3 +13,9 @@ dependencies {
 
   annotationProcessor(libs.lombok)
 }
+
+tasks.test {
+  // Enable parallel test execution
+  systemProperty("junit.jupiter.execution.parallel.enabled", true)
+  systemProperty("junit.jupiter.execution.parallel.mode.default", "concurrent")
```
Comment (philipliu) [inline api-schema/build.gradle.kts:20]: Should we also set `junit.jupiter.execution.parallel.mode.classes.default = concurrent` so the test classes can run in parallel?

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt [hunk 4]
```diff
@@ -4,38 +4,34 @@ import io.mockk.*
...
 import org.junit.jupiter.params.ParameterizedTest
 import org.junit.jupiter.params.provider.EnumSource
-import org.stellar.anchor.auth.ApiAuthJwt.CallbackAuthJwt
-import org.stellar.anchor.auth.ApiAuthJwt.CustodyAuthJwt
-import org.stellar.anchor.auth.ApiAuthJwt.PlatformAuthJwt
+import org.stellar.anchor.auth.ApiAuthJwt.*
 import org.stellar.anchor.auth.AuthType.*
 import org.stellar.anchor.util.AuthHeader
 
+@Execution(SAME_THREAD)
```
Comment (philipliu) [inline core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt:16]: Does this override `junit.jupiter.execution.parallel.mode.default`?

**Review by philipliu — COMMENTED** [review]
Were you seeing data races related to `Sep10Challenge`? I'm wondering why we need a synchronized wrapper.

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt (hunk 4 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt:16]: Yes it does.

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt [hunk 5]
```diff
@@ -1076,31 +1052,4 @@ class Sep31ServiceTest {
...
-
-    // exception maybe thrown for invalid input
-    every { clientsConfig.getClientConfigBySigningKey("Invalid Public Key") } returns null
-    if (!shouldThrowExceptionWithInvalidInput) {
-      val clientNameNotFound = sep31Service.getClientName("Invalid Public Key")
-      assertNull(clientNameNotFound)
-    } else {
-      assertThrows<BadRequestException> { sep31Service.getClientName("Invalid Public Key") }
-    }
-  }
```
Comment (JiahuiWho) [inline core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt:1105]: Is this test removed? Or is it merge/rebase conflict?

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt (hunk 2 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt]: Removed `@Disabled`

File: core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt (hunk 5 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/sep31/Sep31ServiceTest.kt]: Added back.

File: api-schema/build.gradle.kts [hunk 6]
```diff
@@ -13,3 +13,26 @@ dependencies {
 
   annotationProcessor(libs.lombok)
 }
+
+tasks.test {
+  // Enable parallel test execution
+  systemProperty("junit.jupiter.execution.parallel.enabled", false)
```
Comment (philipliu) [inline api-schema/build.gradle.kts:19]: Should this be `true`?

File: core/build.gradle.kts [hunk 7]
```diff
@@ -116,3 +132,27 @@ publishing {
   apply<SigningPlugin>()
   configure<SigningExtension> { sign(publishing.publications) }
 }
+
+// TODO: when we enable parallelization for all sub-projects, we can extract the following block.
+tasks.test {
```
Comment (philipliu) [inline core/build.gradle.kts:136]: Isn't this already set above?

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt [hunk 8]
```diff
@@ -4,38 +4,34 @@ import io.mockk.*
...
-    unmockkAll()
-  }
-
   @ParameterizedTest
   @EnumSource(AuthType::class)
   fun `test AuthHeader creation based on the AuthType`(authType: AuthType) {
     when (authType) {
       JWT -> {
         // Mock calendar to guarantee the jwt token format
+        mockkStatic(Calendar::class)
```
Comment (philipliu) [inline core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt:26]: Can we create a ticket to remove all static mocks from the tests?

File: core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt [hunk 9]
```diff
@@ -22,7 +24,8 @@ import org.stellar.anchor.config.SecretConfig
 import org.stellar.anchor.filter.Sep10JwtFilter.APPLICATION_JSON_VALUE
 import org.stellar.anchor.filter.Sep10JwtFilter.JWT_TOKEN
 
-@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+@Execution(SAME_THREAD)
+@Order(85)
```
Comment (philipliu) [inline core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt:25]: How do we decide the order when we add new tests?

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt [hunk 10]
```diff
@@ -69,6 +65,8 @@ class AuthHelperTest {
         val wantCustodyAuthHeader =
           AuthHeader("Authorization", "Bearer ${jwtService.encode(wantCustodyJwt)}")
         assertEquals(wantCustodyAuthHeader, gotCustodyAuthHeader)
+        unmockkAll()
+        unmockkStatic(Calendar::class)
```
Comment (philipliu) [inline core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt:66]: I thought `unmockkStatic` and `mockkAll` were problematic?

File: platform/build.gradle.kts (hunk 1 above)
Comment (philipliu) [reply platform/build.gradle.kts:60]: Should `junit.jupiter.execution.parallel.enabled` be `false` then?

Comment (lijamie98) [reply platform/build.gradle.kts:60]: Yes. Set it to false.

File: api-schema/build.gradle.kts (hunk 6 above)
Comment (lijamie98) [reply api-schema/build.gradle.kts:19]: Yes. Set to `true` now.

File: core/build.gradle.kts (hunk 7 above)
Comment (lijamie98) [reply core/build.gradle.kts:136]: Duplicated code. Cleaned up and removed duplicated one.

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt (hunk 8 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt:26]: https://stellarorg.atlassian.net/browse/ANCHOR-488

File: core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt (hunk 9 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt:25]: If the test is not long, don't assign the order. I made the slower test with higher priority. 

May be we can make it soemthing like `@Order(1000 - $seconds_to_run)`

File: core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt (hunk 10 above)
Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/auth/AuthHelperTest.kt:66]: The unmockStatic only unmock one class. It is better than `clearAll()`. When we removed `mockkStatic(Calender::class)` in `@SetUp`, we can remove this.

`unmockkAll()` is a mistake. Removed.

File: core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt (hunk 9 above)
Comment (philipliu) [reply core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt:25]: Are you seeing a noticeable improvement by ordering the tests? I think if we aren't seeing it, we should not be manually ordering them.

Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt:25]: It does improve if the thread count is slow.

Comment (lijamie98) [reply core/src/test/kotlin/org/stellar/anchor/filter/Sep10JwtFilterTest.kt:25]: By default, the order should not be assigned unless it is taking long to run.

**Review by philipliu — APPROVED** [review]

**Pull request merged**